### PR TITLE
Bump Loki.

### PR DESCRIPTION
**Pull request by jeschkies**
Summary
- Migrates `ring`, `kv` and `flagext` to `dskit`.
- Imports overrides from Cortex and Thanos.
- Enables `cortextools` to lint LogQL `pattern`.

### CODE REVIEW

**Comment by jeschkies** [comment]
This is required to make https://github.com/grafana/deployment_tools/pull/19340 pass CI.

File: .errcheck-exclude [hunk 1]
```diff
@@ -1,2 +1,3 @@
 (github.com/mitchellh/colorstring).Println
 (github.com/go-kit/kit/log.Logger).Log
+(github.com/go-kit/log.Logger).Log
```
Comment (jeschkies) [inline .errcheck-exclude:3]: This was supposed to fix
```
Error: Error return value of `(github.com/go-kit/log.Logger).Log` is not checked (errcheck)
```
not sure why it's still failing :shrug:

File: pkg/bench/ring_check.go [hunk 2]
```diff
@@ -53,15 +54,25 @@ func NewRingChecker(id string, instanceName string, cfg RingCheckConfig, workloa
...
 	}
-	cfg.MemberlistKV.MetricsRegisterer = prometheus.DefaultRegisterer
+	reg := prometheus.DefaultRegisterer
+	cfg.MemberlistKV.MetricsRegisterer = reg
 	cfg.MemberlistKV.Codecs = []codec.Codec{
 		ring.GetCodec(),
 	}
-	r.MemberlistKV = memberlist.NewKVInitService(&cfg.MemberlistKV, logger)
+
+	dnsProviderReg := prometheus.WrapRegistererWithPrefix(
```
Comment (dannykopping) [inline pkg/bench/ring_check.go:63]: Does this change the names of the existing metrics?

File: go.mod [hunk 3]
```diff
@@ -3,17 +3,20 @@ module github.com/grafana/cortex-tools
 go 1.16
 
 require (
-	cloud.google.com/go/bigtable v1.2.0
+	cloud.google.com/go/bigtable v1.3.0
```
Comment (dannykopping) [inline go.mod:6]: Did you do a general `go get -u`, or are these changes related to the revendoring of Loki?

Comment (jeschkies) [reply go.mod:6]: I've only done a `go get github.com/grafana/loki@5b9f5b9efaa52bfa737e963392975e942e83861b` Ah. Should I have done `-u`?

File: pkg/bench/ring_check.go (hunk 2 above)
Comment (jeschkies) [reply pkg/bench/ring_check.go:63]: The [original](https://github.com/cortexproject/cortex/blob/v1.9.0/pkg/ring/kv/memberlist/kv_init_service.go#L42) has no Prometheus metrix. This is just copy&paste from cortex. Should I use `reg` instead?

File: go.mod (hunk 3 above)
Comment (dannykopping) [reply go.mod:6]: Nope, that's good :+1:

File: pkg/bench/ring_check.go (hunk 2 above)
Comment (dannykopping) [reply pkg/bench/ring_check.go:63]: I have no idea - the authors will probably have to weigh in on this

Comment (jeschkies) [reply pkg/bench/ring_check.go:63]: @replay you fingerprints are on the `ring_check.go` and @aknuds1 you've migrated Cortex to dskit for `NewKVInitService`. Should we use a prefix or just go with the `prometheus.DefaultRegisterer`here?

Comment (aknuds1) [reply pkg/bench/ring_check.go:63]: @jeschkies in order to retain the `cortex_` metrics prefix, you have to configure the same prefix here yes.

Comment (replay) [reply pkg/bench/ring_check.go:63]: I'm not very familiar with this code, I show up in the git blame only because I renamed some symbols. I'd like to defer to somebody who is more familiar

Comment (jeschkies) [reply pkg/bench/ring_check.go:63]: > you do want to retain the metrics prefix from before porting to dskit right?

Yes. So I think we should go with this solution then.

@replay thanks for taking a look nonetheless. 

@dannykopping I think this is good to go then :slightly_smiling_face:

File: pkg/bench/ring_check.go [hunk 4]
```diff
@@ -53,15 +54,25 @@ func NewRingChecker(id string, instanceName string, cfg RingCheckConfig, workloa
...
+			reg,
+		),
+	)
+	dnsProvider := dns.NewProvider(logger, dnsProviderReg, dns.GolangResolverType)
+	r.MemberlistKV = memberlist.NewKVInitService(&cfg.MemberlistKV, logger, dnsProvider, reg)
 	cfg.RingConfig.KVStore.MemberlistKV = r.MemberlistKV.GetMemberlistKV
 
 	var err error
-	r.Ring, err = ring.New(cfg.RingConfig, "ingester", ring.IngesterRingKey, prometheus.DefaultRegisterer)
+	r.Ring, err = ring.New(cfg.RingConfig, "ingester", ring.IngesterRingKey, logger, reg)
```
[the remark this answers is not in the document]
Comment (jeschkies) [reply pkg/bench/ring_check.go:75]: This is the ring though and I just pass the logger. As for `NewKVInitSerive` I will leave it and pass `dnsProviderReg` as it's done in Cortex.

File: pkg/bench/ring_check.go (hunk 2 above)
Comment (dannykopping) [reply pkg/bench/ring_check.go:63]: I believe @gotjosh is one of the main authors here - it'd be good to get their opinion on this

File: pkg/bench/ring_check.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/bench/ring_check.go:75]: @jeschkies ugh sorry, wrong place for the comment.

File: pkg/bench/ring_check.go [hunk 5]
```diff
@@ -53,15 +54,25 @@ func NewRingChecker(id string, instanceName string, cfg RingCheckConfig, workloa
...
+
+	dnsProviderReg := prometheus.WrapRegistererWithPrefix(
+		"cortex_",
+		prometheus.WrapRegistererWith(
+			prometheus.Labels{"name": "memberlist"},
+			reg,
+		),
+	)
+	dnsProvider := dns.NewProvider(logger, dnsProviderReg, dns.GolangResolverType)
+	r.MemberlistKV = memberlist.NewKVInitService(&cfg.MemberlistKV, logger, dnsProvider, reg)
```
Comment (aknuds1) [inline pkg/bench/ring_check.go:71]: @jeschkies I am unsure if reg should be passed to NewKVInitService without the cortex_ prefix.

File: pkg/bench/ring_check.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/bench/ring_check.go:75]: @jeschkies I've updated my comment since my concern is the same for `ring.New`.

File: pkg/bench/ring_check.go (hunk 5 above)
Comment (jeschkies) [reply pkg/bench/ring_check.go:71]: Ah. Got it.

File: pkg/bench/ring_check.go [hunk 6]
```diff
@@ -53,15 +54,25 @@ func NewRingChecker(id string, instanceName string, cfg RingCheckConfig, workloa
...
+
+	dnsProviderReg := prometheus.WrapRegistererWithPrefix(
+		"cortex_",
+		prometheus.WrapRegistererWith(
+			prometheus.Labels{"name": "memberlist"},
+			reg,
+		),
+	)
+	dnsProvider := dns.NewProvider(logger, dnsProviderReg, dns.GolangResolverType)
+	r.MemberlistKV = memberlist.NewKVInitService(&cfg.MemberlistKV, logger, dnsProvider, dnsProviderReg)
```
Comment (aknuds1) [inline pkg/bench/ring_check.go:71]: Sorry about the confusion from before, but @pstibrany pointed out that we haven't historically prefixed memberlist metrics with `cortex_`, so it shouldn't be done here either most likely.
```suggestion
	r.MemberlistKV = memberlist.NewKVInitService(&cfg.MemberlistKV, logger, dnsProvider, reg)
```

File: pkg/bench/ring_check.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/bench/ring_check.go:75]: Never mind, @pstibrany confirmed we don't historically use `cortex_` as prefix for memberlist metrics.

**Review by aknuds1 — COMMENTED** [review]
Updated my comments regarding using `cortex_` as prefix for memberlist metrics, based on feedback from @pstibrany.

**Review by aknuds1 — APPROVED** [review]
LGTM
